### DISCUSSION TITLE
What is the difference between headers set in axiosInstance.default.headers and in axiosInstance.default.headers.common?

### DISCUSSION THREAD

**Discussion by LeonardThomasSenacor** [axios/axios#6601] [Q&A]
While setting an Accept header for an AxiosInstance, I noticed that when the headers are passed initially via `const instance = axios.create({headers: {Accept: 'application/json'}})`, the header will be set under `instance.defaults.headers.Accept`. 

But wherever I read about overwriting default headers of axios it always mentions overwriting the header under `instance.defaults.headers.common` (see also https://www.npmjs.com/package/axios#custom-instance-defaults). Moreover, whether I call `axios.create({headers: {Accept: 'application/json'}})` or `axios.create()`, the value of `instance.defaults.headers.common.Accept` will be 'application/json, text/plain, */*', so the headers passed during create do not influence what's under common. I can only manually overwrite it with `instance.defaults.headers.common = {Accept: 'application/json'}`

So which one takes precedence? `instance.defaults.headers.Accept` or `instance.defaults.headers.common.Accept`? I know that if I set it for example directly for `instance.defaults.headers.post.Accept`, then this one will take precedence because it is more specific.

Or is this a bug in axios 1.7.7 and `instance.defaults.headers.Accept` should not even be possible?

**Comment by harshagarwalnyu** [comment]
The difference is where in the merge pipeline each one lives.

When axios makes a request, it merges headers in this order (each step can override the previous):

1. `defaults.headers.common` - sent on every method (GET, POST, PUT, DELETE, etc.)
2. `defaults.headers[method]` - sent only for that HTTP method (e.g. `defaults.headers.post`)
3. Per-request headers you pass directly

When you do `axios.create({headers: {Accept: '...'}})`, that `Accept` goes into a slot that behaves like step 1 but is separate from `defaults.headers.common`. This is why the two values don't interfere with each other - they're different inputs to the merge.

In practice: use `defaults.headers.common` for things like `Authorization` that should go on every request regardless of method. Use the method-specific slots for things that only make sense on certain methods (like `Content-Type` on POST/PUT). The create-time headers are useful for defaults scoped to a specific client instance.

```js
const instance = axios.create({ baseURL: 'https://api.example.com' })

// these two are equivalent for "send on all methods":
instance.defaults.headers.common['Authorization'] = `Bearer ${token}`
// vs
instance.defaults.headers['Authorization'] = `Bearer ${token}`

// method-specific:
instance.defaults.headers.post['Content-Type'] = 'application/json'
```

The `common` key is just the conventional name axios uses internally - it's not magic, it's just processed first in the merge chain.
